docs/connect.html

```
@@ -543,6 +543,67 @@ <h6>ACL requirements</h6>
         </tbody>
     </table>
 
+    <h4><a id="connect_plugindiscovery" href="#connect_plugindiscovery">Plugin Discovery</a></h4>
+
+    <p>Plugin discovery is the name for the strategy which the Connect worker uses to find plugin classes and make them accessible to configure and run in Connectors and Tasks. This is controlled by the <a href="#connectconfigs_plugin.discovery"><code>plugin.discovery</code> worker configuration</a>, and has a significant impact on worker startup time. <code>SERVICE_LOAD</code> is the fastest strategy, but care should be taken to verify that plugins are compatible before setting this configuration to <code>SERVICE_LOAD</code>.</p>
+
+    <p>Prior to version 3.6, this strategy was not configurable, and behaved like the <code>ONLY_SCAN</code> mode which is compatible with all plugins. For version 3.6 and later, this mode defaults to <code>HYBRID_WARN</code> which is also compatible with all plugins, but logs a warning for all plugins which are incompatible with the other modes. For unit-test environments that use the <code>EmbeddedConnectCluster</code> this defaults to the <code>HYBRID_FAIL</code> strategy, which stops the worker with an error if an incompatible plugin is detected. Finally, the <code>SERVICE_LOAD</code> strategy will silently hide incompatible plugins and make them unusable.</p>
```

I don't think we should explicitly call out the `EmbeddedConnectCluster` class, since it's not public API and discussing it in our docs may mistakenly give people the impression that it is.

Perhaps we could mention the general utility of the `HYBRID_FAIL` mode in testing environments without providing specific details of how we use it ourselves?

---

It's a little unclear what "the other modes" are in "For version 3.6 and later, this mode... logs a warning for all plugins which are incompatible with the other modes."

Maybe "logs a warning for all plugins that cannot be loaded with `SERVICE_LOAD` mode"?

---

Similar thought about "incompatible plugin" in "For unit-test environments... stops the worker with an error if an incompatible plugin is detected."

Maybe "stops the worker if a plugin is detected that cannot be loaded with `SERVICE_LOAD` mode"?

---

Similar to to previous discussions about couching language regarding "unusable" plugins, could we change "Finally, the <code>SERVICE_LOAD</code> strategy will silently hide incompatible plugins and make them unusable" to "Finally, the <code>SERVICE_LOAD</code> strategy will silently hide incompatible plugins and may make them unusable"?

---

Are we being descriptive enough with the term "incompatible plugins" in "Finally, the `SERVICE_LOAD` strategy will silently hide incompatible plugins and make them unusable"? I know we go into more detail later, but if we're going to use that term so frequently, I think we should provide at least a brief summary of what that means in this section. Alternatively, we could say something like "Finally, the `SERVICE_LOAD` strategy will perform plugin scanning faster than other modes, but any plugins that have not been migrated to work with this mode will be silently hidden and may not be usable", which provides a tiny bit more detail and also reiterates the advantages of the `SERVICE_LOAD` mode.

---

> It's a little unclear what "the other modes" are in "For version 3.6 and later, this mode... logs a warning for all plugins which are incompatible with the other modes."

I was saying "incompatible with the other modes" because technically plugins without manifests cause problems in both the hybrid_fail and service_load, and are incompatible with both. I think that including "incompatible with service_load" is less correct but less confusing.

---

> Alternatively, we could say something like "Finally, the SERVICE_LOAD strategy will perform plugin scanning faster than other modes, but any plugins that have not been migrated to work with this mode will be silently hidden and may not be usable", which provides a tiny bit more detail and also reiterates the advantages of the SERVICE_LOAD mode.

Some of this is covered in the end of the previous paragraph as a lead-in to this one; I changed this a bit, but want to avoid repeating myself so I kept the performance discussion in the first paragraph, and added more specifics about incompatible plugins to this paragraph.

---

> I think that including "incompatible with service_load" is less correct but less confusing.

My take is that in general, it's okay to provide information that is not comprehensive but, within a limited scope, accurate. The idea is to only cover scenarios that are highly relevant to users. For example: yes, it's technically correct that plugins without manifests are incompatible with `hybrid_warn`, but since the ultimate goal of this feature is to progress towards `service_load`, it's okay if we only provide information with that context in mind.

---

> Some of this is covered in the end of the previous paragraph as a lead-in to this one; I changed this a bit, but want to avoid repeating myself so I kept the performance discussion in the first paragraph, and added more specifics about incompatible plugins to this paragraph.

The issue with this is that it make it seem like the `service_load` mode only comes with downsides. IMO it's fine to repeat the benefits as long as we're brief. How does "Finally, although it offers the best performance, the `service_load` strategy will hide incompatible plugins..." sound?

---

I've rewritten this to make the previous paragraph more terse, and to explain a bit more in this paragraph about why the service_load strategy is faster.

---

LGTM, thanks!